sharedbuffers/mapped_struct.py

```diff
@@ -342,6 +342,194 @@ def unpack_from(cls, buf, offs, idmap = None):
+        nprobe += 1
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    cython.declare(
+        index_table=tuple,
+        kvpairs=proxied_list,
+        mask=int,
+        nelems=int)
+
+    def __init__(self, mask, nelems, index_table, kvpairs):
+        self.mask = mask
+        self.nelems = nelems
+        self.index_table = index_table
+        self.kvpairs = kvpairs
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        """
+        The binary format of a proxied dictionary is as follows:
+        | size           | type            | attribute name
+        |  8             | long            | mask
+        |  8             | long            | nelems
+        | $mask + 1 * 16 | array of long's | index_table
+        | $nelems        | proxied_list    | kvpairs
+
+        In other words, there are 2 tables: an index vector and a key/value
+        array. The index vector contains the positions in the second array,
+        and their hash values, to speed up comparisons. Unused buckets have
+        a position of -1. The key/value array is packed into a proxied list.
+        """
+
+        xlen = len(obj)
+        size = _upsize(xlen + xlen / 2)   # Aim for a load factor of ~50%
+        indexes = [-1, 0] * size
+        kvpairs = []
+        nelem = 0
+
+        for key, val in obj.iteritems():
+            _add_kv(indexes, kvpairs, key, val)
+            nelem += 1
+
+        len_m1 = len(indexes) / 2 - 1
+        struct.pack_into("=QQ", buf, offs, len_m1, nelem)
+        offs += 16
+        abuf = buffer(array.array('l', indexes))
+        buf[offs:offs + len(abuf)] = abuf
+        offs += len(abuf)
+
+        return offs + mapped_tuple.pack_into(kvpairs, buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        mask, nelems = struct.unpack_from("=QQ", buf, offs)
+        offs += 16
+        idx_size = (mask + 1) * 16
+        index_table = array.array('l')
+        index_table.fromstring(buffer(buf[offs:offs + idx_size]))
+        index_table = tuple(index_table)
+        offs += idx_size
+        kvpairs = proxied_list.unpack_from(buf, offs, idmap)
+        return proxied_dict(mask, nelems, index_table, kvpairs)
+
+    @cython.locals(hval=cython.long, index=cython.long, nprobe=cython.long,
+        i2=cython.long, pos=cython.long)
+    def get_aux(self, key, default_val, out):
+        """
+        The lookup algorithm computes a position by doing hash(key) modulo len(index_vector).
+        Since the index vector is always rounded to a power of 2, this is a
+        very fast operation.
+
+        In addition, before comparing keys, we do a quick check against the
+        saved hash value. Because the same restrictions as with Python apply,
+        said operation is stable, and must match if the keys compare equal.
+
+        Collisions are resolved with a combination of open addressing and
+        linear probing.
```

Comment:
You have 2 implementations of this. You really should have only one.

I'd propose you use `_search_hkey` (one of the variants) for this. It has better worst-case performance than that of a hash table (which is O(N)), and even lower than O(log N) most of the time, with O(1) disk I/Os, which means it's a good fit for memory-mapped files.

The on-disk format is also more compact.

Reply:
Doesn't this limit us to only numeric keys or values ?

Reply:
No. Hashes are always numeric. Values are item offsets.

Reply:
Se how StringIdMapper does its thing. You'd do something similar:

Pack a matrix of:
```
[ hash0  key_off   val_off ]
...
[ hashN  key_off   val_off ]
```

Pack each key/val and populate the matrix, then sort the matrix by hash. You can do all that with numpy efficiently.

Then on reading `search_hkey` will help you get the first matching hash, then you do a sequential search to find the matching key (should be a small run since they'd all be hash collisions).

All of that is in StringIdMapper. You can factor out the common bits and reuse them (careful not to hurt the mapper's performance in the process, use Cython annotations properly).

Reply:
My problem comes with the sequential search part, because search_hkey returns the found value (or the default, if not present), instead of, say, an index (which it can't return). In that case, the lookup is downgraded to a linear search.

Reply:
Did you read `_search_hkey`?